### DISCUSSION TITLE
v2.0.0

### DISCUSSION THREAD

**Discussion by cxxxr** [lem-project/lem#626] [General]
## SDL2 Frontend

The highlight of this release is the implementation of the SDL2 frontend. This feature adds mouse support, graphic capabilities, and Windows support.

## Graphics
Thanks to SDL2's graphics feature, it is now possible to display images and shapes. You can draw at any location on a specified buffer or window.
![screenshot-1](https://github.com/lem-project/lem/assets/13656378/3d9ee4ca-abfc-4981-9837-162842100468)

### Mouse Support

I've added support for mouse clicks and scrolling. Moreover, inserting S-expressions at the cursor position with a middle mouse click, as well as selection by double-click or triple-click, has been made possible.

### Addition of Color Themes

Until now, the default themes were limited to emacs-dark and emacs-light. However, with this release, I've added over 180 new themes. These themes were created by @lukpank.
https://github.com/lukpank/lem-base16-themes

Additionally, I've added the `M-x list-color-themes` command, which allows you to display and select from the list of available themes.

![color-theme](https://github.com/lem-project/lem/assets/13656378/58d56545-8421-43ef-b4b8-67d37da941b3)

Furthermore, the selected theme is saved as a plist in $HOME/.lem/config.lisp and applied when the editor is restarted.

## UI Improvement

I have significantly improved the default look of lem. I've changed the look & feel of elements such as floating-window borders, frame-multiplexer tabs, and the mode line. Moreover, I've added highlighting to the line at the cursor position.

I've also revamped the UI for listing buffers (C-x C-b) and displaying the slime connection list (M-x lisp-connection-list).

![list-buffers](https://github.com/lem-project/lem/assets/13656378/48b1b7d4-9824-4cea-ba8b-9ae0d7314dc4)

Furthermore, I've updated the design of the completion window that appears with commands like C-x C-f.

### grep

I've improved the UI for displaying multiple candidates when using `grep` or `M-.`

![screenshot-grep](https://github.com/lem-project/lem/assets/13656378/2e6cf604-238d-47de-853d-be43ba02ab5f)

### directory-mode

I've improved the display when opening directories using M-x Find File(C-x C-f). Major updates include:

- Display of emojis corresponding to files
- Added mouse support
- Ability to open files not supported by lem with external programs

![screenshot-directory-mode](https://github.com/lem-project/lem/assets/13656378/8c790229-1f7c-4581-b093-9c56d4d85420)

## Asynchronous Completion

I've made completion calculations asynchronous. This improvement reduces lag during completion calculations.

## Lisp Mode

Several minor improvements have been made:

- Indent adjustment when yanking
- Addition of a feature that does not insert ')' when there's no corresponding opening parenthesis
- Adjustment of output position to REPL
- Hovering over underlined parts during function/file compilation pops up messages

![compile-error](https://github.com/lem-project/lem/assets/13656378/e5b756a8-5a95-4bfe-8c59-7516845e570e)

## vi-mode

Many bug fixes and refactorings for vi-mode have been contributed by @theangelperalta.
![vi-mode](https://github.com/lem-project/lem/assets/13656378/9f9e9c13-9278-4bac-90e0-262de057a050)

## Other Changes

There have been numerous other changes, including:

- Add var and defun for removing trailing whitespace on save #608
- Close socket gracefully when `slime-quit' #622
- Addition of json-mode
- Minor changes and large-scale refactoring, such as replacing internal modules used in lsp-mode, language-server
- A significant number of bug fixes

## Known Issues

There are several unresolved issues related to SDL2:

- The SDL2 frontend may not work with certain graphic drivers, especially when installing graphic drivers on Linux.
- On Linux, the system may become unresponsive when returning from suspend.
- Modifier key changes in macOS settings or Linux's xmodmap may not be reflected.

<hr /><em>This discussion was created from the release <a href='https://github.com/lem-project/lem/releases/tag/v2.0.0'>v2.0.0</a>.</em>

**Comment by tmheath** [comment]
Hello, I'm new... There were a couple things that I was going to look at trying to do from the onboarding stuff, but I feel like I've been wasting my time looking for where the windows launcher cmd script is. Can someone explain how the general release process works? It's not really a bugfix but a huge QoL improvement via adding (cd %~dp0\c\n) to the script just before the launch statement. The behavior seems to switch from crashing when launched from anywhere outside the downloaded folder to being able to be launched from anywhere. An additional thing I might look at is using the start command to see if it can work from a new terminal/process instead of consuming the calling terminal.

**Comment by cxxxr** [comment]
@tmheath 

Thanks for looking and sorry for the late reply.
I use this repository for releases.
And lem.bat is here.
https://github.com/lem-project/lem-deploy/blob/main/windows/lem.bat

For windows deployments, I prepare a set of executables at https://github.com/lem-project/lem-deploy/blob/main/build-windows-sdl2.lisp, zip them up and upload them.
